### PR TITLE
Model according to the plan

### PR DESCRIPTION
**Pull request by dominik271828**
462563710_1662771491321167_6221772392089628635_n
There is a mistake concerning the diagram, there shouldn't be an arrow from ID in product pointing to the ACCOUNT_PRODUCT_GROUP. It should rather point to the ACCOUNT table.

### CODE REVIEW

File: API/users/models.py [hunk 1]
```diff
@@ -1,2 +1,45 @@
...
+
+class Category(models.Model):
+    name = models.CharField(max_length=64)
+    parent = models.ForeignKey('self', null=True, on_delete=models.SET_NULL)
+    def __str__(self):
+        return self.name
+
+class ProductGroup(models.Model):
+    owner = models.ForeignKey(User, on_delete=models.CASCADE)
+    account = models.ManyToManyField(Account)
```
Comment (AdamBalski) [inline API/users/models.py:25]:
```suggestion
    accounts = models.ManyToManyField(Account)
```

File: API/users/models.py [hunk 2]
```diff
@@ -1,2 +1,45 @@
...
+
+class Product(models.Model):
+    name = models.CharField(max_length=64)
+    category = models.ForeignKey(Category, null=True, on_delete=models.SET_NULL)
+    group = models.ForeignKey(ProductGroup, on_delete = models.CASCADE)
+    def __str__(self):
+        return self.name
+
+class TransactionItem(models.Model):
+    quantity = models.FloatField()
```
Comment (kapiw04) [inline API/users/models.py:39]: shouldn't it be `IntegerField`?

File: API/users/models.py [hunk 3]
```diff
@@ -1,2 +1,45 @@
...
+    price = models.FloatField()
+    product = models.ForeignKey(Product, on_delete=models.CASCADE)
```
Comment (kapiw04) [inline API/users/models.py:41]: I am not sure about `CASCADE` here. I can imagine a scenario where `Product` record gets deleted by the user. In this case we would probably like to keep `TransactionItem` in database. Maybe change it to `SET_NULL`?

File: API/users/models.py [hunk 4]
```diff
@@ -1,2 +1,45 @@
...
+    name = models.CharField(max_length=64)
+    parent = models.ForeignKey('self', null=True, on_delete=models.SET_NULL)
+    def __str__(self):
+        return self.name
+
+class ProductGroup(models.Model):
+    owner = models.ForeignKey(User, on_delete=models.CASCADE)
+    account = models.ManyToManyField(Account)
+    def __str__(self):
+        return self.name
```
Comment (kapiw04) [inline API/users/models.py:29]: There's no `name` field in this model

File: API/users/models.py [hunk 5]
```diff
@@ -1,2 +1,45 @@
...
+    name = models.CharField(max_length=64)
+    user = models.ForeignKey(User, on_delete=models.CASCADE)
+    def __str__(self):
+        return self.name
+
+class Transaction(models.Model):
+    account = models.ForeignKey(Account, on_delete=models.CASCADE)
+    timestamp = models.DateTimeField()
+    def __str__(self):
+        return self.name
```
Comment (kapiw04) [inline API/users/models.py:17]: No `name` field in this model

File: API/users/models.py [hunk 6]
```diff
@@ -1,2 +1,45 @@
...
+
+class TransactionItem(models.Model):
+    quantity = models.FloatField()
+    price = models.FloatField()
+    product = models.ForeignKey(Product, on_delete=models.CASCADE)
+    transaction = models.ForeignKey(Transaction, on_delete=models.CASCADE)
+    def __str__(self):
+        return self.name
```
Comment (kapiw04) [inline API/users/models.py:44]: Again, no name field ;)

**Review by kapiw04 — CHANGES_REQUESTED** [review]
Generally LGTM. See minor improvements I suggested. When you're done, maybe run 
```sh
python3 manage.py makemigrations
python3 manage.py migrate
```
to finalise this feature

File: API/users/models.py (hunk 2 above)
Comment (AdamBalski) [reply API/users/models.py:33]: I think that is for cases such as buying 1.5kg of potates or anything by weight

File: API/users/models.py (hunk 5 above)
Comment (dominik271828) [reply API/users/models.py:17]: I'm removing the `def __str(self)` wherever there isn't a possibility to include one then

File: API/users/models.py [hunk 7]
```diff
@@ -1,2 +1,46 @@
...
+from django.contrib.auth.models import User
 
 # Create your models here.
+
+class Account(models.Model):
+    # account_set added implicitly
+    name = models.CharField(max_length=64)
+    user = models.ForeignKey(User, on_delete=models.CASCADE)
+    def __str__(self):
+        return self.name
```
Comment (AdamBalski) [inline API/users/models.py:11]: idk whether we need `__str__` on all classes, but if we need, maybe let's make the result more detailed?
sth like this
```suggestion
        return f"Account(name: '{self.name}', user_id:'{self.user_id}')"
```
?

**Review by AdamBalski — CHANGES_REQUESTED** [review]

File: API/users/models.py (hunk 7 above)
Comment (kapiw04) [reply API/users/models.py]: I think we can leave `__str__` as it is. More detailed info can be returned by `__repr__`. That being said, probably we won't need any of those lol

File: API/users/models.py [hunk 8]
```diff
@@ -1,2 +1,38 @@
...
 
 # Create your models here.
+
+class Account(models.Model):
+    # account_set added implicitly
+    name = models.CharField(max_length=64)
+    user = models.ForeignKey(User, on_delete=models.CASCADE)
+    def __str__(self):
+        return f"Account(name: '{self.name}', user_id:'{self.user_id}')"
+
```
Comment (AdamBalski) [inline API/users/models.py:12]:
```suggestion

```

File: API/users/models.py [hunk 9]
```diff
@@ -1,2 +1,38 @@
...
+class Transaction(models.Model):
+    account = models.ForeignKey(Account, on_delete=models.CASCADE)
+    timestamp = models.DateTimeField()
+
+class Category(models.Model):
+    name = models.CharField(max_length=64)
+    parent = models.ForeignKey('self', null=True, on_delete=models.SET_NULL)
+    def __str__(self):
+        return self.name
+
```
Comment (AdamBalski) [inline API/users/models.py:22]:
```suggestion

```

File: API/users/models.py [hunk 10]
```diff
@@ -1,2 +1,38 @@
...
+class Product(models.Model):
+    name = models.CharField(max_length=64)
+    category = models.ForeignKey(Category, null=True, on_delete=models.SET_NULL)
+    group = models.ForeignKey(ProductGroup, on_delete = models.CASCADE)
+
+class TransactionItem(models.Model):
+    quantity = models.FloatField()
+    price = models.FloatField()
+    product = models.ForeignKey(Product, on_delete=models.CASCADE)
+    transaction = models.ForeignKey(Transaction, null=True, on_delete=models.SET_NULL)
```
Comment (AdamBalski) [inline API/users/models.py:36]:
```suggestion
    product = models.ForeignKey(Product, on_delete=models.SET_NULL)
    transaction = models.ForeignKey(Transaction, null=True, on_delete=models.CASCADE)
```

File: API/users/models.py [hunk 11]
```diff
@@ -1,2 +1,38 @@
+from django.db import models
+from django.contrib.auth.models import User
 
 # Create your models here.
+
+class Account(models.Model):
+    # account_set added implicitly
```
Comment (AdamBalski) [inline API/users/models.py:7]: are you talking about product group set?

Comment (dominik271828) [reply API/users/models.py:7]: yes, sorry

**Review by AdamBalski — APPROVED** [review]

**Review by kapiw04 — APPROVED** [review]

**Pull request merged**